### PR TITLE
Issue #3837: report lambda with no braces when contents span multiple lines

### PR DESCRIPTION
**Pull request by strkkk**
#3837
Fix + changes to resolve violations in checkstyle code.
Reg report https://strkkk.github.io/3837/

### CODE REVIEW

**Comment by rnveach** [comment]
codecov is not in the CI list.

**Comment by strkkk** [comment]
> Please, add something like that to the docs.

It would make docs less readable, isn't it? :)

**Comment by romani** [comment]
@strkkk ,

Please add to inputs more complicated cases, example https://strkkk.github.io/3837/spring-framework/index.html#A1
Inner lambdas are good cases to inputs.

>Please, add something like that to the docs.

Please add to xdoc.

**Comment by pbludov** [comment]
> > Please, add something like that to the docs.

Something like this:

Rationale: it is hard to read lambdas that span multiple lines, for example:
 ```
  allowedFuture.addCallback(result -> assertEquals("Invalid response",
    EnumSet.of(HttpMethod.GET, HttpMethod.OPTIONS, HttpMethod.HEAD), result),
    ex -> fail(ex.getMessage()));
```
versus
```
allowedFuture.addCallback(result -> {
  return assertEquals("Invalid response",
    EnumSet.of(HttpMethod.GET, HttpMethod.OPTIONS, HttpMethod.HEAD), result);
  },
  ex -> fail(ex.getMessage()));
```

**Comment by strkkk** [comment]
@pbludov in what section this example should be placed, in your opinion?
This check checks different tokens and lambda is only one of them. "Rationale" section should describe idea in general, but this is only case for lambda.

> Please add to inputs more complicated cases, example https://strkkk.github.io/3837/spring-framework/index.html#A1
> Inner lambdas are good cases to inputs.

@romani I added test case

**Comment by romani** [comment]
Please add example to https://checkstyle.org/config_blocks.html#NeedBraces_Examples

**Comment by pbludov** [comment]
> @pbludov in what section this example should be placed, in your opinion?
@strkkk, you are right. A fix for the `rationale` section is out of the scope of this issue. I'll start a new one.

**Comment by strkkk** [comment]
@romani @pbludov 
Please decide, is this needed or not :)

> A fix for the `rationale` section is out of the scope of this issue
> Please add example to https://checkstyle.org/config_blocks.html#NeedBraces_Examples

**Comment by romani** [comment]
There is no Rationale section at all in doc, no need to introduce it.

Please add example for lambda, `To configure the check for ...`

**Comment by strkkk** [comment]
@romani @pbludov I added examples, please take a look

**Comment by romani** [comment]
@rnveach , please finalize.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java [hunk 1]
```diff
@@ -416,8 +439,22 @@ private static boolean isSingleLineIf(DetailAST literalIf) {
...
-        return isOnSameLine(lambda, block);
+        final DetailAST lastLambdaToken = getLastLambdaToken(lambda);
+        return isOnSameLine(lambda, lastLambdaToken);
+    }
+
+    /**
+     * Looks for last lambda token.
+     *
+     * @param lambda token to check.
+     * @return last lambda token
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java]: `last lambda token` => `last token in lambda`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java [hunk 2]
```diff
@@ -416,8 +439,22 @@ private static boolean isSingleLineIf(DetailAST literalIf) {
...
+     */
+    private static DetailAST getLastLambdaToken(DetailAST lambda) {
+        DetailAST node = lambda.getLastChild();
+        while (node.getLastChild() != null) {
+            node = node.getLastChild();
+        }
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java]: Would the code be cleaner with
````
DetailAST node = lambda;
do {
  node = node.getLastChild();
} while (node.getLastChild() != null);
````
?

File: src/xdocs/config_blocks.xml [hunk 3]
```diff
@@ -859,6 +859,29 @@ switch (num) {
...
+            Results in following:
+          </p>
+          <pre>
+allowedFuture.addCallback(result -> assertEquals("Invalid response",
+  EnumSet.of(HttpMethod.GET, HttpMethod.OPTIONS), result), // violation, lambda spans 2 lines
+  ex -> fail(ex.getMessage())); // OK
+
+allowedFuture.addCallback(result -> {
+  return assertEquals("Invalid response",
+    EnumSet.of(HttpMethod.GET, HttpMethod.OPTIONS), result); // OK
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/xdocs/config_blocks.xml]: changed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java [hunk 4]
```diff
@@ -416,8 +439,22 @@ private static boolean isSingleLineIf(DetailAST literalIf) {
...
      */
     private static boolean isSingleLineLambda(DetailAST lambda) {
-        final DetailAST block = lambda.getLastChild();
-        return isOnSameLine(lambda, block);
+        final DetailAST lastLambdaToken = getLastLambdaToken(lambda);
+        return isOnSameLine(lambda, lastLambdaToken);
+    }
+
+    /**
+     * Looks for last lambda token.
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java (hunk 1 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java]: looks like it, corrected

**Comment by romani** [comment]
Please rebase on latest master to fix Travis build if it fail.

**Comment by strkkk** [comment]
It failed (I rebased), but reason is not log overflow but some other infra issue
> apt-get install failed
272The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install xsltproc xmlstarlet" failed and exited with 100 during .

**Pull request merged**